### PR TITLE
Add CachedTypeAhead Recipe

### PR DESCRIPTION
**Pull request by azizhk**
For now, keeping things simple and easy to understand using promises.
Ideally, this should be done using Suspense.
Suspense should handle cancellation of Promises in case of unmounts as well.
Hoping for the best.

### CODE REVIEW

File: src/components/TypeAhead.tsx [hunk 1]
```diff
@@ -102,4 +110,61 @@ class TypeAhead extends React.PureComponent<TypeaheadProps, TypeaheadState> {
...
+      return this.setState({ query });
+    }
+
+    const optionsPromise = this.props.children(query);
+    this.setState({
+      query: query,
+      cache: {
+        ...this.state.cache,
+        [query]: {
+          promise: optionsPromise
```
Comment (ritz078) [inline src/components/TypeAhead.tsx]: why save promise when it is never used from state?

File: src/components/TypeAhead.tsx [hunk 2]
```diff
@@ -102,4 +110,61 @@ class TypeAhead extends React.PureComponent<TypeaheadProps, TypeaheadState> {
   }
 }
 
-export default TypeAhead;
+export class CachedTypeAhead extends React.PureComponent<
```
Comment (ritz078) [inline src/components/TypeAhead.tsx]: named exports are not available from outside by default. 2 options here - make this `TypeAhead.Cached` or export explicitly as a new component. I would prefer the first one.

**Comment by ritz078** [comment]
tests should be fixed by taking a pull.

**Comment by ritz078** [comment]
Bug - Even if there is no result, the dropdown shows.

File: src/components/typings/Typeahead.ts [hunk 3]
```diff
@@ -33,3 +33,18 @@ export interface TypeaheadState {
   value: string;
   showSuggestions: boolean;
 }
+
+type keys = Exclude<keyof TypeaheadProps, "onChange" | "children">;
+export type CachedTypeAheadProps = Pick<TypeaheadProps, keys> & {
```
Comment (ritz078) [inline src/components/typings/Typeahead.ts]: the whole code uses `interface` wherever possible.

File: src/components/TypeAhead.tsx (hunk 1 above)
Comment (ritz078) [reply src/components/TypeAhead.tsx]: also currently there is a flicker when new data is fetched. the drop-down clears and then when data is fetched, it comes back as for some period of time there are no results. I don't think that is the expected behavior?

File: src/components/typings/Typeahead.ts (hunk 3 above)
Comment (ritz078) [reply src/components/typings/Typeahead.ts]: fixed by my commit

File: src/components/TypeAhead.tsx (hunk 1 above)
Comment (azizhk) [reply src/components/TypeAhead.tsx]: Was initially hoping to cancel the Promise on character change, but realized that it's difficult to expose cancellation token while keeping functionality simple (i.e. children function returns Promise to `<Option />`s)
Need to do some more research before finalize API.

File: src/components/typings/Typeahead.ts (hunk 3 above)
Comment (azizhk) [reply src/components/typings/Typeahead.ts]: Did not get it? Did you push?

**Comment by azizhk** [comment]
Again now deployment failed because zeit's now uses npm instead of yarn and the latest version of react-spring is shipped again with broken types.
https://github.com/react-spring/react-spring/issues/520#issuecomment-463045726

Could not find anything on how to use yarn instead of npm on Zeit's documentation https://zeit.co/docs/v2/deployments/configuration

react-spring is moving to Typescript 🥇 btw so these problems should get solved.

**Pull request merged**